File: LeetCodeAlgos/Python/sumOfLeftLeaves.py
# Given the root of a binary tree, return the sum of all left leaves.
# A leaf is a node with no children. A left leaf is a leaf that is the left child of another node.

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        res = 0
        if not root: return res
        if not root.left and not root.right: return root.val

        first, next = [root], []
        while first:
            for n in first:
                if n.left:
                    x = n.left
                    if not x.left and not x.right:
                        res += n.left.val
                    else:
                        next.append(n.left)
                if n.right: 
                    next.append(n.right)
            first, next = next, []
        return res

s = Solution()
print(s.sumOfLeftLeaves(TreeNode( 3,  TreeNode( 9 ), TreeNode( 20, TreeNode( 15 ), TreeNode( 7 )))))
print(s.sumOfLeftLeaves(TreeNode( 3 )))

File: LeetCodeAlgos/Java/SumOfLeftLeaves.java
// Given the root of a binary tree, return the sum of all left leaves.

// A leaf is a node with no children. A left leaf is a leaf that is the left child of another node.
class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;
  TreeNode() {}
  TreeNode(int val) { this.val = val; }
  TreeNode(int val, TreeNode left, TreeNode right) {
      this.val = val;
      this.left = left;
      this.right = right;
  }
}

class Solution {
  public int sumOfLeftLeaves(TreeNode root) {
    // Rule out no root or leaves
    if (root == null) {
      return 0;
    }
    return dp(root, false, 0);
  }

  private int dp(TreeNode node, boolean left, int sum) {
    // If a leaf and on left
    if ((node.left == null && node.right == null) && left) {
      sum += node.val;
    } else {
      // If left subs,  sum all left leaves
      sum += node.left != null ? dp(node.left, true, 0) : 0;
      // If right subs, sum all left leaves
      sum += node.right != null ? dp(node.right, false, 0) : 0;
    }
    // return sum
    return sum;
  }
}

class SumOfLeftLeaves {
  public static void main(String[] args) {
    Solution s = new Solution();

    TreeNode root1 = new TreeNode(3,
      new TreeNode(9),
      new TreeNode(20,
        new TreeNode(15),
        new TreeNode(7)
      )
    );
    testSumOfLeftLeaves(1, root1, 24, s);

    TreeNode root2 = new TreeNode(1);
    testSumOfLeftLeaves(2, root2, 0, s);
  }

  public static void testSumOfLeftLeaves(int testNum, TreeNode root, int expected, Solution s) {
    int result = s.sumOfLeftLeaves(root);
    System.out.println("Test " + testNum + ": " + result);
    assert (result == expected);
  }
}